### PR TITLE
[dashboard] Show prebuild status based on prebuild workspace record

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Currently, the Prebuild detail view (with logs and status) uses the `WorkspaceInstance` status. This is inconsistent with respect to the Prebuild List view. This PR uses the `prebuild` status instead of status from the `WorkspaceInstance`.

This PR does not remove all instances of this inconcistent usage (another in [ConfigureProject.tsx](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/projects/ConfigureProject.tsx#L368)). It will be removed in a subsequent PR.

## Related Issue(s)

* Related to #5611
* Relates to #7384
* Relates to #8197

## How to test

1. [Preview env](https://mp-pb-status.staging.gitpod-dev.com/)
2. Add/clone https://gitlab.com/gitpod-milan/gitpod-large-image as a Project
3. Trigger builds for the following branches:
	* main
	* pass
	* timeout
	* fail-prebuild
4. Observe various statuses reported
5. Trigger a new prebuild and cancel it manually - observe cancelled status

## Release Notes

```release-note
Dashboard reports prebuild status consistently between list view and detail view.
```

## How it looks
Screenshot 2022-03-15 at 19 41 18
Screenshot 2022-03-15 at 19 41 11
Screenshot 2022-03-15 at 19 41 02
Screenshot 2022-03-15 at 19 42 51
Screenshot 2022-03-15 at 19 42 39
Screenshot 2022-03-15 at 19 45 31

## Open questions
* Designs could be improved with respect to showing the underlying error. I think we can follow-up with a cleanup of the styling to communicate the underlying error message more cleanly

/hold

- [x] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by easyCZ** [comment]
@gtsiolis The current way we report the underlying error (next to the status) is a bit ugly. Do you have thoughts on what we could do here? For now, we only have a string as the error but I'm working on making this a bit richer as well.

Co-related: The statuses we have here don't always convey well. For example, we show `Failed` in red, but it was in a way a "user error" - something defined by the prebuild tasks didn't complete succesfully. I think here it would helpful to better convey that the system is fine, but the task is not.

**Comment by gtsiolis** [comment]
[ETA](https://en.wikipedia.org/wiki/Estimated_time_of_arrival): Adding this to [scheduled](https://github.com/orgs/gitpod-io/projects/29) to take a look later today or tomorrow.

Thanks for the ping @easyCZ! 🏓

**Comment by andrew-farries** [comment]
Gave this a quick look just now and after triggering prebuilds on all branches, they all failed with:

image

**Comment by easyCZ** [comment]
@andrew-farries I've rebased and re-deployed this branch, I've also run the above tests and I've been able to reproduce the desired changes.

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 1]
```diff
@@ -329,12 +329,66 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
...
+            </div>
+            <div className="flex space-x-1 items-center text-gray-400">
+                <PrebuildStatusDescription prebuild={prebuild} />
+            </div>
+        </div>
+    );
+}
+
+// Deprecated. Use PrebuildStatus instead.
+export function PrebuildInstanceStatus(props: { prebuildInstance?: WorkspaceInstance; prebuild?: PrebuildWithStatus }) {
```
Comment (AlexTugarev) [inline components/dashboard/src/projects/Prebuilds.tsx:391]: I see, it's still used in ConfigureProject component. OK

**Review by AlexTugarev — DISMISSED** [review]
Works as advertised!

Thanks!

File: components/dashboard/src/projects/Prebuild.tsx [hunk 2]
```diff
@@ -167,7 +167,7 @@ export default function () {
                         />
                     </div>
                     <div className="h-20 px-6 bg-gray-50 dark:bg-gray-800 border-t border-gray-200 dark:border-gray-600 flex space-x-2">
-                        {prebuildInstance && <PrebuildInstanceStatus prebuildInstance={prebuildInstance} />}
+                        {prebuildInstance && <PrebuildStatus prebuild={prebuild!} />}
```
Comment (jankeromnes) [inline components/dashboard/src/projects/Prebuild.tsx:170]: Shouldn't this be:

```suggestion
                        {prebuild && <PrebuildStatus prebuild={prebuild} />}
```

**Review by jankeromnes — DISMISSED** [review]
This is fantastic, many thanks @easyCZ! 👏

Code changes also look good to me, and the PR seems to work as advertised. 🎯

@gtsiolis Would you like to review the "How it looks" screenshots in the [PR description](https://github.com/gitpod-io/gitpod/pull/8805#issue-1168614444)? 😁 (Personally, this looks good to me and is already better than what we have today in production, so we could also merge this and make any adjustments in follow-ups. 💭)

File: components/dashboard/src/projects/Prebuild.tsx (hunk 2 above)
Comment (AlexTugarev) [reply components/dashboard/src/projects/Prebuild.tsx:170]: Indeed. I wonder, why we use the `WorkspaceInstance` at all in `PrebuildLogs`. The call site can use `onPrebuildUpdate` instead, so we even can get rid of a subsequent API call 👍🏻

**Comment by gtsiolis** [comment]
Taking a look now! 👀 

But feel free to merge this as is.

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 3]
```diff
@@ -329,12 +329,66 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
...
+            return (
+                <span>
+                    Prebuild timed out. Either the image, or the prebuild tasks took too long. {props.prebuild.error}
+                </span>
+            );
+        case "available":
+            if (props.prebuild?.error) {
+                return (
+                    <span>
+                        The tasks executed in the prebuild returned a non-zero exit code. {props.prebuild.error}
```
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:357]: **question**: This could be confusing. Could we use simpler words to describe the error. Maybe something like the following?

> A task configured to run in the prebuild returned a non-zero exit code.

Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:357]: **suggestion**: Also, the error appended there is already included in the log output above, right? If yes, I think we could safely drop it from there. WDYT?

Frame 369

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 4]
```diff
@@ -329,12 +329,66 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
...
+    switch (props.prebuild.status) {
+        case "queued":
+            return <span>Prebuild is queued and will be processed when there is execution capacity.</span>;
+        case "building":
+            return <span>Prebuild is currently in progress.</span>;
+        case "aborted":
+            return (
+                <span>
+                    Prebuild has been cancelled. Either a user cancelled it, or the prebuild rate limit has been
+                    exceeded. {props.prebuild.error}
```
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:342]: **suggestion**: Again, a simpler message here could help while dropping appending information that is already duplicated. WDYT?

> Prebuild has been manually cancelled or prebuild rate limit has been reached.

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 5]
```diff
@@ -329,12 +329,66 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
     }
 }
 
+function PrebuildStatusDescription(props: { prebuild: PrebuildWithStatus }) {
+    switch (props.prebuild.status) {
+        case "queued":
+            return <span>Prebuild is queued and will be processed when there is execution capacity.</span>;
```
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:335]: **thought**: Sounds accurate but a bit scary for humans, no? Maybe we could use something like the following:

> Prebuild is queued and will be processed in a moment.

File: components/dashboard/src/projects/Prebuilds.tsx [hunk 6]
```diff
@@ -329,12 +329,66 @@ export function prebuildStatusIcon(prebuild?: PrebuildWithStatus) {
...
+                    Prebuild has been cancelled. Either a user cancelled it, or the prebuild rate limit has been
+                    exceeded. {props.prebuild.error}
+                </span>
+            );
+        case "failed":
+            return <span>Prebuild failed for system reasons. Please contact support. {props.prebuild.error}</span>;
+        case "timeout":
+            return (
+                <span>
+                    Prebuild timed out. Either the image, or the prebuild tasks took too long. {props.prebuild.error}
```
Comment (gtsiolis) [inline components/dashboard/src/projects/Prebuilds.tsx:350]: **suggestion**: Maybe the following suffices?

> Prebuild timed out because the image or prebuild tasks took too long.

Optiontally, we could append some information about the max prebuild limit.

> Prebuild timed out because the image or prebuild tasks took too long. Maximum prebuild duration is 1 hour.

**Review by gtsiolis — COMMENTED** [review]
I'm late here but left some comments we could have addressed in this iteration. Posted here in case we'd like to open follow up issues or PRs for these.

Cc @andrew-farries @jankeromnes @AlexTugarev @easyCZ

**Comment by easyCZ** [comment]
> I'm late here but left some comments we could have addressed in this iteration. Posted here in case we'd like to open follow up issues or PRs for these.
> 
> Cc @andrew-farries @jankeromnes @AlexTugarev @easyCZ

Thanks for the feedback. That makes sense. From my perspective, I'd be interested in exposing the full error message but possibly it can be inside (by default) collapsed component. Surfacing the information does allow us to get better feedback on the errors from users and be able to address them. WDYT @gtsiolis?

My intention for follow-up PRs is to clean-up the errors the server provides such that we treat them as "publicly facing errors" and include actionable feedback.

> Optiontally, we could append some information about the max prebuild limit.

On the topic of these, the error message should really come directly from the backend. We're entering a world where workspace features can configurable and in that case only the backend will be authoritative over what the timeout (or other properties) actually are. I think we'll see us move away more from the UI having the error text and more it being sourced directly from the DB as a user facing error.

File: components/dashboard/src/projects/Prebuilds.tsx (hunk 3 above)
Comment (easyCZ) [reply components/dashboard/src/projects/Prebuilds.tsx:357]: Definitely. I'm trying to clean it up incrementally as there's a lot of tech debt with respect where we collect the "truth" from. Unify first, then improve the messaging. 

I'll cut a ticket for this concrete case to not show the same error twice.

**Comment by gtsiolis** [comment]
> From my perspective, I'd be interested in exposing the full error message but possibly it can be inside (by default) collapsed component. Surfacing the information does allow us to get better feedback on the errors from users and be able to address them. WDYT @gtsiolis?

Fully agree on beng able to see the full error message. However, we can mask the form of the error on the elements that do not scale well with long error messages and use the full error message on areas that scale better like the logs output.

For example, currently the time out error easily breaks out of the one sentence limit and also introduces long message that is hard to parse and includes duplicate information (see 01h00m).

🍊 🍊 🍊 🍊 

> My intention for follow-up PRs is to clean-up the errors the server provides such that we treat them as "publicly facing errors" and include actionable feedback.

Looking forward for this one, @easyCZ! Feel free to loop me in if you need any more feedback on this.

🍋 🍋 🍋 🍋 

> On the topic of these, the error message should really come directly from the backend.

Also, fully agree. But still masking the type of the error or including a short version of the error message could help to include in small areas like here or in tooltips could be useful.

> I think we'll see us move away more from the UI having the error text and more it being sourced directly from the DB as a user facing error.

Sounds good! 🙏

File: components/dashboard/src/projects/Prebuilds.tsx (hunk 3 above)
Comment (gtsiolis) [reply components/dashboard/src/projects/Prebuilds.tsx:357]: > Unify first, then improve the messaging.

💯 

> I'll cut a ticket for this concrete case to not show the same error twice.

Thanks, @easyCZ! 🏀

**Comment by easyCZ** [comment]
> Fully agree on beng able to see the full error message. However, we can mask the form of the error on the elements that do not scale well with long error messages and use the full error message on areas that scale better like the logs output.

**Comment by gtsiolis** [comment]
Now that this is merged, shall we close any of the related issues linked in the PR description?

- https://github.com/gitpod-io/gitpod/issues/5611
- https://github.com/gitpod-io/gitpod/issues/7384
- https://github.com/gitpod-io/gitpod/issues/8197

Cc @easyCZ @jankeromnes @AlexTugarev

**Comment by easyCZ** [comment]
Thanks for the follow-up @gtsiolis. I believe we can action it as follows:
* https://github.com/gitpod-io/gitpod/issues/5611 - close, this is indeed resolved by this PR
* https://github.com/gitpod-io/gitpod/issues/7384 - already closed
* https://github.com/gitpod-io/gitpod/issues/8197 - there remains one more instance of this issue on the Project Settings view. I'm tackling this in https://github.com/gitpod-io/gitpod/pull/8831 and will close once this lands.

I'll go and close the first now.

**Comment by gtsiolis** [comment]
Thanks, @easyCZ! 🙏

File: components/dashboard/src/projects/Prebuilds.tsx (hunk 3 above)
Comment (gtsiolis) [reply components/dashboard/src/projects/Prebuilds.tsx:357]: **fyi**: I've opened https://github.com/gitpod-io/gitpod/issues/9111 to keep track of this discussion and the unresolved comments below.
